### PR TITLE
[PROF-4904] New Stack collector is feature-complete

### PR DESCRIPTION
**Pull request by ivoanjo**
The new stack collector introduced in #1966 gets feature-complete in this PR.

By feature-complete I mean:

* There is extensive test coverage with many corner cases being checked
* Our custom `rb_profile_frames` got quite a lot of fixes to exactly match the Ruby reference stack trace api (`Thread#backtrace_locations`)
* It works for all supported Rubies, including 2.1 and 2.2 which were a though nut to crack
* It matches the behavior of the old profiling codepaths, e.g. it has placeholders for threads with empty stacks that are running in native code and for threads with too deep stacks
* etc...

I don't quite like how long and how complex this PR got. Thanks for putting up with it 😅 , and feel free to challenge me if this should be further broken down. Reviewing it commit-by-commit may make more sense, helping to show what bits were added why.

Note: This PR builds on top of #1966.
Note 2: I'm aware that specs for "Datadog::Profiling::Collectors::Stack when sampling a thread with a stack that is deeper than the configured max_frames" are failing. I missed this previously as it did not happen on macOS (works fine on macOS). I'll push a fix for it ASAP, but otherwise this is ready for reviewing.

### CODE REVIEW

**Comment by ivoanjo** [comment]
Ooops mistakenly forgot to open this as a draft!

File: ext/ddtrace_profiling_native_extension/stack_recorder.h [hunk 1]
```diff
@@ -22,7 +22,7 @@
 #define   ALLOC_SPACE_VALUE {.type_ = VALUE_STRING("alloc-space"),   .unit = VALUE_STRING("bytes")}
 #define    HEAP_SPACE_VALUE {.type_ = VALUE_STRING("heap-space"),    .unit = VALUE_STRING("bytes")}
 
-const static ddprof_ffi_ValueType enabled_value_types[] = {CPU_TIME_VALUE, CPU_SAMPLES_VALUE, WALL_TIME_VALUE};
+static const ddprof_ffi_ValueType enabled_value_types[] = {CPU_TIME_VALUE, CPU_SAMPLES_VALUE, WALL_TIME_VALUE};
```
Comment (marcotc) [inline ext/ddtrace_profiling_native_extension/stack_recorder.h:25]: Found the most important change right here!

File: spec/datadog/profiling/collectors/interesting_backtrace_helper.rb [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  def hello
+    IbhClassWithStaticMethod.hello
+  end
+end
+
+class IbhClassC
+  include IbhModuleD
+end
+
+$ibh_a_proc = proc { IbhClassC.new.hello }
```
Comment (marcotc) [inline spec/datadog/profiling/collectors/interesting_backtrace_helper.rb:47]: Globals! How about slapping all these `$` bad boys into a module in this file, like `IbhGlobal`?

Comment (marcotc) [reply spec/datadog/profiling/collectors/interesting_backtrace_helper.rb:47]: To be clear there, my suggestion will likely accomplish the same level of code isolation, but I just find it jarring to see these `$` around me. It makes me slightly confused when I glance a random `$` while reading code here.

File: ext/ddtrace_profiling_native_extension/collectors_stack.c [hunk 3]
```diff
@@ -56,50 +73,215 @@ static VALUE _native_sample(VALUE self, VALUE thread, VALUE recorder_instance, V
...
+
+  // Idea: Should we release the global vm lock (GVL) after we get the data from `rb_profile_frames`? That way other Ruby threads
+  // could continue making progress while the sample was ingested into the profile.
+  //
+  // Other things to take into consideration if we go in that direction:
+  // * Is it safe to call `rb_profile_frame_...` methods on things from the `stack_buffer` without the GVL acquired?
+  // * We need to make `VALUE` references in the `stack_buffer` visible to the Ruby GC
+  // * Should we move this into a different thread entirely?
+  // * If we don't move it into a different thread, does releasing the GVL on a Ruby thread mean that we're introducing
+  //   a new thread switch point where there previously was none?
```
[the remark this answers is not in the document]
Comment (ivoanjo) [reply ext/ddtrace_profiling_native_extension/collectors_stack.c:112]: Indeed. Part of the awkwardness of releasing the GVL or even sending it to a different thread that operates without the GVL is that we'd need to duplicate a lot of the Ruby-level stuff, because it requires the GVL being held to access it. That's annoying because it would introduce an extra copy; tracking of a bunch of heap-allocated stuff, etc.

There's been some discussion with the other profiler libraries about possibly changing libddprof/libdatadog to expose its string table. I've been thinking it could be used to "solve" this issue -- e.g. we could copy all the Ruby strings into the string table (with the GVL being held), avoiding any extra copies, and then we'd have a lot less data to send to a background thread.

So yeah, definitely I'll be keeping an eye on this one.

File: ext/ddtrace_profiling_native_extension/collectors_stack.c [hunk 4]
```diff
@@ -56,50 +73,215 @@ static VALUE _native_sample(VALUE self, VALUE thread, VALUE recorder_instance, V
...
+      .labels = labels,
+    }
+  );
+}
+
+sampling_buffer *sampling_buffer_new(unsigned int max_frames) {
+  if (max_frames < 5) rb_raise(rb_eArgError, "Invalid max_frames: value must be >= 5");
+  if (max_frames > MAX_FRAMES_LIMIT) rb_raise(rb_eArgError, "Invalid max_frames: value must be <= " MAX_FRAMES_LIMIT_AS_STRING);
+
+  sampling_buffer* buffer = xcalloc(1, sizeof(sampling_buffer));
```
[the remark this answers is not in the document]
Comment (ivoanjo) [reply ext/ddtrace_profiling_native_extension/collectors_stack.c:250]: Actually thanks for calling me out on this!

`xcalloc` is indeed provided by Ruby -- it's aliased to `ruby_xcalloc`:
```
 * Allocates a  storage instance.  It is  largely the same as  system malloc(),
 * except:
 *
 *   - It raises Ruby exceptions instead of returning NULL, and
 *   - In case of `ENOMEM` it tries to GC to make some room.
```

https://github.com/ruby/ruby/blob/0293f8ca8e1fbe8e4202c2d211bb11f4e0ffa5fd/include/ruby/internal/xmalloc.h#L53-L147

The part I had missed thus far is this one:

```
 * @note       It doesn't return NULL.
```

Because Ruby will call `ruby_memerror` aka try to raise our friend the `NoMemoryError` exception in case of `NULL`. So I've been adding more `NULL`-checks than needed. I'll do a pass an clean-up any such unneeded checks :)

File: ext/ddtrace_profiling_native_extension/collectors_stack.c [hunk 5]
```diff
@@ -56,50 +73,215 @@ static VALUE _native_sample(VALUE self, VALUE thread, VALUE recorder_instance, V
...
+  if (buffer == NULL) rb_memerror();
```
[the remark this answers is not in the document]
Comment (ivoanjo) [reply ext/ddtrace_profiling_native_extension/collectors_stack.c:251]: You're 💯% right, will remove :)

File: ext/ddtrace_profiling_native_extension/stack_recorder.h (hunk 1 above)
Comment (ivoanjo) [reply ext/ddtrace_profiling_native_extension/stack_recorder.h:25]: :hurtrealbad:

File: spec/datadog/profiling/collectors/interesting_backtrace_helper.rb (hunk 2 above)
Comment (ivoanjo) [reply spec/datadog/profiling/collectors/interesting_backtrace_helper.rb]: Fixed in ce1c971c07012ed2d3c5cee92930417a4b1abd36

File: ext/ddtrace_profiling_native_extension/collectors_stack.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivoanjo) [reply ext/ddtrace_profiling_native_extension/collectors_stack.c]: Fixed in 68e4bc50b5423f1db365916d0599eaeb5204d285

File: ext/ddtrace_profiling_native_extension/collectors_stack.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivoanjo) [reply ext/ddtrace_profiling_native_extension/collectors_stack.c]: Fixed in 68e4bc50b5423f1db365916d0599eaeb5204d285

**Comment by ivoanjo** [comment]
One last rebase to solve a last-minute conflict with #1966 and this PR is now on top of master!

Just need to chase down that CI failure relating to max_frames before merging :)
